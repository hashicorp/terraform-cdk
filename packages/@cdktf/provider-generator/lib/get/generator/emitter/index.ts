// Copyright (c) HashiCorp, Inc
// SPDX-License-Identifier: MPL-2.0
export * from "./attributes-emitter";
export * from "./resource-emitter";
export * from "./struct-emitter";
