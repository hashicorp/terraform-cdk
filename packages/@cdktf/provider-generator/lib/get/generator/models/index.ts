// Copyright (c) HashiCorp, Inc
// SPDX-License-Identifier: MPL-2.0
export * from "./attribute-model";
export * from "./attribute-type-model";
export * from "./resource-model";
export * from "./scope";
export * from "./struct";
