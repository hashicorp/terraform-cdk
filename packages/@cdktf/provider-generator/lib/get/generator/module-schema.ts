// Copyright (c) HashiCorp, Inc
// Licensed under the Mozilla Public License 2.0
export interface ModuleSchema {
  name: string;
  empty?: boolean;
  inputs: Input[];
  outputs: Output[];
}

export interface Output {
  name: string;
  type: string;
}

export interface Input {
  name: string;
  type: string;
  description?: string;
  default?: string;
  required: boolean;
}
