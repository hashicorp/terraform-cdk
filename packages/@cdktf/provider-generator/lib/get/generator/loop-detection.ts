// Copyright (c) HashiCorp, Inc
// SPDX-License-Identifier: MPL-2.0
import { AttributeModel } from "./models/attribute-model";

type AttributeIdentifier = [string, Set<string>];
const equalAttributeIdentifiers = (
  x: AttributeIdentifier,
  y: AttributeIdentifier
) =>
  x[0] === y[0] &&
  x[1].size === y[1].size &&
  [...x[1]].every((x) => y[1].has(x));

function typeStructure(model: AttributeModel) {
  if (!model.type.isComplex) {
    return model.type.storedClassType;
  }

  return `<complex:{${model.type.struct?.attributes
    .map((att) => att.name)
    .sort()}}${model.type.typeModelType}>`;
}

function getAttributeIdentifier(model: AttributeModel): AttributeIdentifier {
  return [
    model.terraformName,
    new Set(
      model.type.struct!.attributes.map((a) => `${a.name}:${typeStructure(a)}`)
    ),
  ];
}

/**
 * Detects recursive attribute schemas and returns an Object
 * describing what property should be linking to which class
 */
export function detectAttributeLoops(attributes: AttributeModel[]): {
  [loopEntrypoint: string]: string; // attribute path to be replaced -> attribute path of the type to replace with
} {
  // We aproximate a class as a set of attributes names
  // This is not a perfect approximation, but it's good enough for our purposes
  const redirects: Record<string, string> = {};

  function depthFirstSearch(
    attribute: AttributeModel,
    path: string[] = [],
    knownStructs: { [attributePath: string]: AttributeIdentifier } = {}
  ) {
    const name = attribute.terraformName;
    const struct = attribute.type.struct;
    if (!struct) {
      return;
    }

    const structIdentifier = getAttributeIdentifier(attribute);
    // Detect if we visited this already
    const visited = Object.entries(knownStructs).find(([, attrIdentifier]) =>
      equalAttributeIdentifiers(structIdentifier, attrIdentifier)
    );

    if (visited) {
      // We have a loop, we don't need to search deeper
      const toBeReplacedWith = visited[0];
      const toReplace = [...path, name].join(".");
      redirects[toReplace] = toBeReplacedWith;
    } else {
      // Search deeper
      struct.attributes.forEach((a) =>
        depthFirstSearch(a, [...path, name], {
          ...knownStructs,
          // We haven't visited this yet, add it to the list of known structs
          [[...path, name].join(".")]: structIdentifier,
        })
      );
    }
  }

  attributes.forEach((attr) => depthFirstSearch(attr));

  return redirects;
}
