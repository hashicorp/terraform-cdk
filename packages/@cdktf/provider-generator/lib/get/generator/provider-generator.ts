// Copyright (c) HashiCorp, Inc
// SPDX-License-Identifier: MPL-2.0
import { CodeMaker, toCamelCase } from "codemaker";
import { LANGUAGES, logger, TerraformProviderConstraint } from "@cdktf/commons";
import {
  FQPN,
  parseFQPN,
  ProviderName,
  ProviderSchema,
} from "./provider-schema";
import { ResourceModel } from "./models";
import { ResourceParser } from "./resource-parser";
import { ResourceEmitter, StructEmitter } from "./emitter";
import {
  ConstructsMakerProviderTarget,
  ConstructsMakerTarget,
} from "../constructs-maker";

interface ProviderData {
  name: string;
  source: string;
  version: string;
}

const isMatching = (
  target: ConstructsMakerTarget,
  terraformSchemaName: string
): boolean => {
  if (target.isModule) return false;

  const elements = terraformSchemaName.split("/");

  if (elements.length === 1) {
    return target.source === terraformSchemaName;
  } else {
    const [hostname, scope, provider] = elements;

    if (!hostname || !scope || !provider) {
      throw new Error(`can't handle ${terraformSchemaName}`);
    }

    return target.name === provider;
  }
};

export interface ProviderConstraints {
  [fqn: string]: ProviderData;
}

export class TerraformProviderGenerator {
  private resourceParser = new ResourceParser();
  private resourceEmitter: ResourceEmitter;
  private structEmitter: StructEmitter;
  public versions: { [fqpn: string]: string | undefined } = {};

  constructor(
    private readonly code: CodeMaker,
    private readonly schema: ProviderSchema
  ) {
    this.code.indentation = 2;
    this.resourceEmitter = new ResourceEmitter(this.code);
    this.structEmitter = new StructEmitter(this.code);
  }

  private getProviderByConstraint(
    providerConstraint: ConstructsMakerTarget
  ): FQPN | undefined {
    return Object.keys(this.schema.provider_schemas || {}).find((fqpn) =>
      isMatching(providerConstraint, fqpn)
    ) as FQPN | undefined;
  }

  public generate(providerConstraint: ConstructsMakerTarget) {
    const fqpn = this.getProviderByConstraint(providerConstraint);
    if (!fqpn) {
      logger.debug(
        `Could not find provider constraint for ${providerConstraint} in schema: ${JSON.stringify(
          this.schema,
          null,
          2
        )}`
      );
      throw new Error(
        `Could not find provider with constraint ${providerConstraint}`
      );
    }

    const providerVersion = this.schema.provider_versions?.[fqpn];
    this.emitProvider(fqpn, providerVersion, providerConstraint);
    this.versions[fqpn] = providerVersion;
  }

  public generateAll() {
    for (const fqpn of Object.keys(this.schema.provider_schemas || {})) {
      this.generate(
        new ConstructsMakerProviderTarget(
          new TerraformProviderConstraint(fqpn),
          LANGUAGES[0]
        )
      );
    }
  }

  public async save(outdir: string) {
    await this.code.save(outdir);
  }

  public buildResourceModels(fqpn: FQPN): ResourceModel[] {
    const provider = this.schema.provider_schemas?.[fqpn];
    if (!provider) {
      throw new Error(`Can not find provider '${fqpn}' in schema`);
    }

    const resources = Object.entries(provider.resource_schemas || {}).map(
      ([type, resource]) =>
        this.resourceParser.parse(fqpn, type, resource, "resource")
    );

    const dataSources = Object.entries(provider.data_source_schemas || {}).map(
      ([type, resource]) =>
        this.resourceParser.parse(fqpn, `data_${type}`, resource, "data_source")
    );

    return ([] as ResourceModel[]).concat(...resources, ...dataSources);
  }

  public getClassNameForResource(terraformType: string) {
    return this.resourceParser.getClassNameForResource(terraformType);
  }

  public getNamespaceNameForResource(terraformType: string) {
    return this.resourceParser.getNamespaceNameForResource(terraformType);
  }

  private emitProvider(
    fqpn: FQPN,
    providerVersion?: string,
    constraint?: ConstructsMakerTarget
  ) {
    const { name } = parseFQPN(fqpn);
    const provider = this.schema.provider_schemas?.[fqpn];
    if (!provider) {
      throw new Error(`Can not find provider '${fqpn}' in schema`);
    }

    const files: string[] = [];
    this.buildResourceModels(fqpn).forEach((resourceModel) => {
      if (constraint) {
        resourceModel.providerVersionConstraint = constraint.version;
        resourceModel.terraformProviderSource = constraint.source;
      }
      resourceModel.providerVersion = providerVersion;

      if (resourceModel.structsRequireSharding) {
        files.push(this.emitResourceWithComplexStruct(resourceModel));
      } else {
        files.push(this.emitResource(resourceModel));
      }
      this.emitResourceReadme(resourceModel);
    });

    if (provider.provider) {
      const providerResource = this.resourceParser.parse(
        fqpn,
        `provider`,
        provider.provider,
        "provider"
      );
      if (constraint) {
        providerResource.providerVersionConstraint = constraint.version;
        providerResource.terraformProviderSource = constraint.source;
      }
      providerResource.providerVersion = providerVersion;
      files.push(this.emitResource(providerResource));
      this.emitResourceReadme(providerResource);
    }

    this.emitIndexFile(name, files);
    this.emitLazyIndexFile(name, files);
  }

  private emitResourceReadme(resource: ResourceModel): void {
    const filePath = `${resource.namespaceFolderPath}/README.md`;
    this.code.openFile(filePath);
    this.code.line(`# \`${resource.terraformType}\``);
    this.code.line();
    const type = resource.isProvider
      ? resource.provider
      : resource.terraformType;
    this.code.line(
      `Refer to the Terraform Registory for docs: [\`${type}\`](${resource.linkToDocs}).`
    );
    this.code.closeFile(filePath);
  }

  private emitIndexFile(provider: ProviderName, files: string[]): void {
    const folder = `providers/${provider}`;
    const filePath = `${folder}/index.ts`;
    this.code.openFile(filePath);
    this.code.line("// generated by cdktf get");
    for (const file of files) {
      const dirName = file.replace(`${folder}/`, "").replace("/index.ts", "");
      this.code.line(
        `export * as ${toCamelCase(dirName)} from './${dirName}';`
      );
    }
    this.code.line();
    this.code.closeFile(filePath);
  }

  private emitLazyIndexFile(provider: ProviderName, files: string[]): void {
    const folder = `providers/${provider}`;
    const filePath = `${folder}/lazy-index.ts`;
    this.code.openFile(filePath);
    this.code.line("// generated by cdktf get");
    for (const file of files) {
      const dirName = file.replace(`${folder}/`, "").replace("/index.ts", "");
      this.code.line(
        `Object.defineProperty(exports, '${toCamelCase(
          dirName
        )}', { get: function () { return require('./${dirName}'); } });`
      );
    }
    this.code.line();
    this.code.closeFile(filePath);
  }

  private emitResource(resource: ResourceModel): string {
    this.code.openFile(resource.filePath);
    this.emitFileHeader(resource);
    this.structEmitter.emit(resource);
    this.resourceEmitter.emit(resource);
    this.code.closeFile(resource.filePath);

    return resource.filePath;
  }

  private emitResourceWithComplexStruct(resource: ResourceModel) {
    const generatedFiles = [];

    // drop the last segment of the filepath
    const filePath = resource.filePath;
    this.code.openFile(filePath);
    this.code.line(`// generated from terraform resource schema`);
    this.code.line();

    if (resource.structsRequireSharding) {
      if (resource.referencedTypes.length > 0) {
        this.code.line(
          `import { ${resource.referencedTypes.join(", \n")}} from './${
            resource.structsFolderName
          }'`
        );
      }

      this.code.line(`export * from './${resource.structsFolderName}'`);

      resource.importStatements.forEach((statement) =>
        this.code.line(statement)
      );

      this.structEmitter.emitInterface(resource, resource.configStruct);
      this.resourceEmitter.emit(resource);

      this.code.closeFile(filePath);

      this.structEmitter.emit(resource);
      generatedFiles.push(resource.fileName);
      generatedFiles.push(resource.structsFolderName);
    } else {
      resource.importStatements.forEach((statement) =>
        this.code.line(statement)
      );

      this.structEmitter.emit(resource);
      this.resourceEmitter.emit(resource);
      this.code.closeFile(filePath);
      generatedFiles.push(resource.fileName);
    }

    return filePath;
  }

  private emitFileHeader(resource: ResourceModel) {
    this.code.line(`// ${resource.linkToDocs}`);
    this.code.line(`// generated from terraform resource schema`);
    this.code.line();
    resource.importStatements.forEach((statement) => this.code.line(statement));
    this.code.line();
    this.code.line("// Configuration");
    this.code.line();
  }
}
