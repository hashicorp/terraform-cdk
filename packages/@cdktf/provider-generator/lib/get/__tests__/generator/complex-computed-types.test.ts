// Copyright (c) HashiCorp, Inc
// SPDX-License-Identifier: MPL-2.0
import * as fs from "fs";
import * as os from "os";
import * as path from "path";
import { TerraformProviderGenerator } from "../../generator/provider-generator";
import { CodeMaker } from "codemaker";

test("generate an acm certifacte resource with complex computed types", async () => {
  const code = new CodeMaker();
  const workdir = fs.mkdtempSync(
    path.join(os.tmpdir(), "complex-computed-types.test")
  );
  const spec = JSON.parse(
    fs.readFileSync(
      path.join(__dirname, "fixtures", "aws_acm_certificate.test.fixture.json"),
      "utf-8"
    )
  );
  new TerraformProviderGenerator(code, spec).generateAll();
  await code.save(workdir);

  const output = fs.readFileSync(
    path.join(workdir, "providers/aws/acm/acm-certificate.ts"),
    "utf-8"
  );
  expect(output).toMatchSnapshot(`acm-certificate`);

  const outputIndex = fs.readFileSync(
    path.join(workdir, "providers/aws/acm/index.ts"),
    "utf-8"
  );
  expect(outputIndex).toMatchSnapshot(`index`);
});
