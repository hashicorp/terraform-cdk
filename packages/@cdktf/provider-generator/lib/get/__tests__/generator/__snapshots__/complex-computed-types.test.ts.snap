// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`generate an acm certifacte resource with complex computed types: acm-certificate 1`] = `
"// https://registry.terraform.io/providers/hashicorp/aws/latest/docs/resources/acm_certificate
// generated from terraform resource schema

import { Construct } from 'constructs';
import * as cdktf from 'cdktf';

// Configuration

export interface AcmCertificateConfig extends cdktf.TerraformMetaArguments {
  /**
  * Docs at Terraform Registry: {@link https://registry.terraform.io/providers/hashicorp/aws/latest/docs/resources/acm_certificate#certificate_authority_arn AcmCertificate#certificate_authority_arn}
  */
  readonly certificateAuthorityArn?: string;
  /**
  * Docs at Terraform Registry: {@link https://registry.terraform.io/providers/hashicorp/aws/latest/docs/resources/acm_certificate#certificate_body AcmCertificate#certificate_body}
  */
  readonly certificateBody?: string;
  /**
  * Docs at Terraform Registry: {@link https://registry.terraform.io/providers/hashicorp/aws/latest/docs/resources/acm_certificate#certificate_chain AcmCertificate#certificate_chain}
  */
  readonly certificateChain?: string;
  /**
  * Docs at Terraform Registry: {@link https://registry.terraform.io/providers/hashicorp/aws/latest/docs/resources/acm_certificate#domain_name AcmCertificate#domain_name}
  */
  readonly domainName?: string;
  /**
  * Docs at Terraform Registry: {@link https://registry.terraform.io/providers/hashicorp/aws/latest/docs/resources/acm_certificate#id AcmCertificate#id}
  *
  * Please be aware that the id field is automatically added to all resources in Terraform providers using a Terraform provider SDK version below 2.
  * If you experience problems setting this value it might not be settable. Please take a look at the provider documentation to ensure it should be settable.
  */
  readonly id?: string;
  /**
  * Docs at Terraform Registry: {@link https://registry.terraform.io/providers/hashicorp/aws/latest/docs/resources/acm_certificate#private_key AcmCertificate#private_key}
  */
  readonly privateKey?: string;
  /**
  * Docs at Terraform Registry: {@link https://registry.terraform.io/providers/hashicorp/aws/latest/docs/resources/acm_certificate#subject_alternative_names AcmCertificate#subject_alternative_names}
  */
  readonly subjectAlternativeNames?: string[];
  /**
  * Docs at Terraform Registry: {@link https://registry.terraform.io/providers/hashicorp/aws/latest/docs/resources/acm_certificate#tags AcmCertificate#tags}
  */
  readonly tags?: { [key: string]: string };
  /**
  * Docs at Terraform Registry: {@link https://registry.terraform.io/providers/hashicorp/aws/latest/docs/resources/acm_certificate#validation_method AcmCertificate#validation_method}
  */
  readonly validationMethod?: string;
  /**
  * options block
  *
  * Docs at Terraform Registry: {@link https://registry.terraform.io/providers/hashicorp/aws/latest/docs/resources/acm_certificate#options AcmCertificate#options}
  */
  readonly options?: AcmCertificateOptions;
}
export interface AcmCertificateDomainValidationOptions {
}

export function acmCertificateDomainValidationOptionsToTerraform(struct?: AcmCertificateDomainValidationOptions): any {
  if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) { return struct; }
  if (cdktf.isComplexElement(struct)) {
    throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
  }
  return {
  }
}

export class AcmCertificateDomainValidationOptionsOutputReference extends cdktf.ComplexObject {
  private isEmptyObject = false;

  /**
  * @param terraformResource The parent resource
  * @param terraformAttribute The attribute on the parent resource this class is referencing
  * @param complexObjectIndex the index of this item in the list
  * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
  */
  public constructor(terraformResource: cdktf.IInterpolatingParent, terraformAttribute: string, complexObjectIndex: number, complexObjectIsFromSet: boolean) {
    super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
  }

  public get internalValue(): AcmCertificateDomainValidationOptions | undefined {
    let hasAnyValues = this.isEmptyObject;
    const internalValueResult: any = {};
    return hasAnyValues ? internalValueResult : undefined;
  }

  public set internalValue(value: AcmCertificateDomainValidationOptions | undefined) {
    if (value === undefined) {
      this.isEmptyObject = false;
    }
    else {
      this.isEmptyObject = Object.keys(value).length === 0;
    }
  }

  // domain_name - computed: true, optional: false, required: false
  public get domainName() {
    return this.getStringAttribute('domain_name');
  }

  // resource_record_name - computed: true, optional: false, required: false
  public get resourceRecordName() {
    return this.getStringAttribute('resource_record_name');
  }

  // resource_record_type - computed: true, optional: false, required: false
  public get resourceRecordType() {
    return this.getStringAttribute('resource_record_type');
  }

  // resource_record_value - computed: true, optional: false, required: false
  public get resourceRecordValue() {
    return this.getStringAttribute('resource_record_value');
  }
}

export class AcmCertificateDomainValidationOptionsList extends cdktf.ComplexList {

  /**
  * @param terraformResource The parent resource
  * @param terraformAttribute The attribute on the parent resource this class is referencing
  * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
  */
  constructor(protected terraformResource: cdktf.IInterpolatingParent, protected terraformAttribute: string, protected wrapsSet: boolean) {
    super(terraformResource, terraformAttribute, wrapsSet)
  }

  /**
  * @param index the index of the item to return
  */
  public get(index: number): AcmCertificateDomainValidationOptionsOutputReference {
    return new AcmCertificateDomainValidationOptionsOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
  }
}
export interface AcmCertificateOptions {
  /**
  * Docs at Terraform Registry: {@link https://registry.terraform.io/providers/hashicorp/aws/latest/docs/resources/acm_certificate#certificate_transparency_logging_preference AcmCertificate#certificate_transparency_logging_preference}
  */
  readonly certificateTransparencyLoggingPreference?: string;
}

export function acmCertificateOptionsToTerraform(struct?: AcmCertificateOptionsOutputReference | AcmCertificateOptions): any {
  if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) { return struct; }
  if (cdktf.isComplexElement(struct)) {
    throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
  }
  return {
    certificate_transparency_logging_preference: cdktf.stringToTerraform(struct!.certificateTransparencyLoggingPreference),
  }
}

export class AcmCertificateOptionsOutputReference extends cdktf.ComplexObject {
  private isEmptyObject = false;

  /**
  * @param terraformResource The parent resource
  * @param terraformAttribute The attribute on the parent resource this class is referencing
  */
  public constructor(terraformResource: cdktf.IInterpolatingParent, terraformAttribute: string) {
    super(terraformResource, terraformAttribute, false, 0);
  }

  public get internalValue(): AcmCertificateOptions | undefined {
    let hasAnyValues = this.isEmptyObject;
    const internalValueResult: any = {};
    if (this._certificateTransparencyLoggingPreference !== undefined) {
      hasAnyValues = true;
      internalValueResult.certificateTransparencyLoggingPreference = this._certificateTransparencyLoggingPreference;
    }
    return hasAnyValues ? internalValueResult : undefined;
  }

  public set internalValue(value: AcmCertificateOptions | undefined) {
    if (value === undefined) {
      this.isEmptyObject = false;
      this._certificateTransparencyLoggingPreference = undefined;
    }
    else {
      this.isEmptyObject = Object.keys(value).length === 0;
      this._certificateTransparencyLoggingPreference = value.certificateTransparencyLoggingPreference;
    }
  }

  // certificate_transparency_logging_preference - computed: false, optional: true, required: false
  private _certificateTransparencyLoggingPreference?: string; 
  public get certificateTransparencyLoggingPreference() {
    return this.getStringAttribute('certificate_transparency_logging_preference');
  }
  public set certificateTransparencyLoggingPreference(value: string) {
    this._certificateTransparencyLoggingPreference = value;
  }
  public resetCertificateTransparencyLoggingPreference() {
    this._certificateTransparencyLoggingPreference = undefined;
  }
  // Temporarily expose input value. Use with caution.
  public get certificateTransparencyLoggingPreferenceInput() {
    return this._certificateTransparencyLoggingPreference;
  }
}

/**
* Represents a {@link https://registry.terraform.io/providers/hashicorp/aws/latest/docs/resources/acm_certificate aws_acm_certificate}
*/
export class AcmCertificate extends cdktf.TerraformResource {

  // =================
  // STATIC PROPERTIES
  // =================
  public static readonly tfResourceType = "aws_acm_certificate";

  // ===========
  // INITIALIZER
  // ===========

  /**
  * Create a new {@link https://registry.terraform.io/providers/hashicorp/aws/latest/docs/resources/acm_certificate aws_acm_certificate} Resource
  *
  * @param scope The scope in which to define this construct
  * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
  * @param options AcmCertificateConfig = {}
  */
  public constructor(scope: Construct, id: string, config: AcmCertificateConfig = {}) {
    super(scope, id, {
      terraformResourceType: 'aws_acm_certificate',
      terraformGeneratorMetadata: {
        providerName: 'aws'
      },
      provider: config.provider,
      dependsOn: config.dependsOn,
      count: config.count,
      lifecycle: config.lifecycle,
      provisioners: config.provisioners,
      connection: config.connection,
      forEach: config.forEach
    });
    this._certificateAuthorityArn = config.certificateAuthorityArn;
    this._certificateBody = config.certificateBody;
    this._certificateChain = config.certificateChain;
    this._domainName = config.domainName;
    this._id = config.id;
    this._privateKey = config.privateKey;
    this._subjectAlternativeNames = config.subjectAlternativeNames;
    this._tags = config.tags;
    this._validationMethod = config.validationMethod;
    this._options.internalValue = config.options;
  }

  // ==========
  // ATTRIBUTES
  // ==========

  // arn - computed: true, optional: false, required: false
  public get arn() {
    return this.getStringAttribute('arn');
  }

  // certificate_authority_arn - computed: false, optional: true, required: false
  private _certificateAuthorityArn?: string; 
  public get certificateAuthorityArn() {
    return this.getStringAttribute('certificate_authority_arn');
  }
  public set certificateAuthorityArn(value: string) {
    this._certificateAuthorityArn = value;
  }
  public resetCertificateAuthorityArn() {
    this._certificateAuthorityArn = undefined;
  }
  // Temporarily expose input value. Use with caution.
  public get certificateAuthorityArnInput() {
    return this._certificateAuthorityArn;
  }

  // certificate_body - computed: false, optional: true, required: false
  private _certificateBody?: string; 
  public get certificateBody() {
    return this.getStringAttribute('certificate_body');
  }
  public set certificateBody(value: string) {
    this._certificateBody = value;
  }
  public resetCertificateBody() {
    this._certificateBody = undefined;
  }
  // Temporarily expose input value. Use with caution.
  public get certificateBodyInput() {
    return this._certificateBody;
  }

  // certificate_chain - computed: false, optional: true, required: false
  private _certificateChain?: string; 
  public get certificateChain() {
    return this.getStringAttribute('certificate_chain');
  }
  public set certificateChain(value: string) {
    this._certificateChain = value;
  }
  public resetCertificateChain() {
    this._certificateChain = undefined;
  }
  // Temporarily expose input value. Use with caution.
  public get certificateChainInput() {
    return this._certificateChain;
  }

  // domain_name - computed: true, optional: true, required: false
  private _domainName?: string; 
  public get domainName() {
    return this.getStringAttribute('domain_name');
  }
  public set domainName(value: string) {
    this._domainName = value;
  }
  public resetDomainName() {
    this._domainName = undefined;
  }
  // Temporarily expose input value. Use with caution.
  public get domainNameInput() {
    return this._domainName;
  }

  // domain_validation_options - computed: true, optional: false, required: false
  private _domainValidationOptions = new AcmCertificateDomainValidationOptionsList(this, "domain_validation_options", false);
  public get domainValidationOptions() {
    return this._domainValidationOptions;
  }

  // id - computed: true, optional: true, required: false
  private _id?: string; 
  public get id() {
    return this.getStringAttribute('id');
  }
  public set id(value: string) {
    this._id = value;
  }
  public resetId() {
    this._id = undefined;
  }
  // Temporarily expose input value. Use with caution.
  public get idInput() {
    return this._id;
  }

  // private_key - computed: false, optional: true, required: false
  private _privateKey?: string; 
  public get privateKey() {
    return this.getStringAttribute('private_key');
  }
  public set privateKey(value: string) {
    this._privateKey = value;
  }
  public resetPrivateKey() {
    this._privateKey = undefined;
  }
  // Temporarily expose input value. Use with caution.
  public get privateKeyInput() {
    return this._privateKey;
  }

  // subject_alternative_names - computed: true, optional: true, required: false
  private _subjectAlternativeNames?: string[]; 
  public get subjectAlternativeNames() {
    return this.getListAttribute('subject_alternative_names');
  }
  public set subjectAlternativeNames(value: string[]) {
    this._subjectAlternativeNames = value;
  }
  public resetSubjectAlternativeNames() {
    this._subjectAlternativeNames = undefined;
  }
  // Temporarily expose input value. Use with caution.
  public get subjectAlternativeNamesInput() {
    return this._subjectAlternativeNames;
  }

  // tags - computed: false, optional: true, required: false
  private _tags?: { [key: string]: string }; 
  public get tags() {
    return this.getStringMapAttribute('tags');
  }
  public set tags(value: { [key: string]: string }) {
    this._tags = value;
  }
  public resetTags() {
    this._tags = undefined;
  }
  // Temporarily expose input value. Use with caution.
  public get tagsInput() {
    return this._tags;
  }

  // validation_emails - computed: true, optional: false, required: false
  public get validationEmails() {
    return this.getListAttribute('validation_emails');
  }

  // validation_method - computed: true, optional: true, required: false
  private _validationMethod?: string; 
  public get validationMethod() {
    return this.getStringAttribute('validation_method');
  }
  public set validationMethod(value: string) {
    this._validationMethod = value;
  }
  public resetValidationMethod() {
    this._validationMethod = undefined;
  }
  // Temporarily expose input value. Use with caution.
  public get validationMethodInput() {
    return this._validationMethod;
  }

  // options - computed: false, optional: true, required: false
  private _options = new AcmCertificateOptionsOutputReference(this, "options");
  public get options() {
    return this._options;
  }
  public putOptions(value: AcmCertificateOptions) {
    this._options.internalValue = value;
  }
  public resetOptions() {
    this._options.internalValue = undefined;
  }
  // Temporarily expose input value. Use with caution.
  public get optionsInput() {
    return this._options.internalValue;
  }

  // =========
  // SYNTHESIS
  // =========

  protected synthesizeAttributes(): { [name: string]: any } {
    return {
      certificate_authority_arn: cdktf.stringToTerraform(this._certificateAuthorityArn),
      certificate_body: cdktf.stringToTerraform(this._certificateBody),
      certificate_chain: cdktf.stringToTerraform(this._certificateChain),
      domain_name: cdktf.stringToTerraform(this._domainName),
      id: cdktf.stringToTerraform(this._id),
      private_key: cdktf.stringToTerraform(this._privateKey),
      subject_alternative_names: cdktf.listMapper(cdktf.stringToTerraform, false)(this._subjectAlternativeNames),
      tags: cdktf.hashMapper(cdktf.stringToTerraform)(this._tags),
      validation_method: cdktf.stringToTerraform(this._validationMethod),
      options: acmCertificateOptionsToTerraform(this._options.internalValue),
    };
  }
}
"
`;
